src/dbnode/encoding/proto/custom_marshal.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"math"
+
+	"github.com/golang/protobuf/proto"
+)
+
+// encoding methods correspond to the scalar types defined in the protobuf 3
+// specificaiton: https://developers.google.com/protocol-buffers/docs/proto3#scalar
+type customFieldMarshaller interface {
+	encFloat64(tag int32, x float64)
+	encFloat32(tag int32, x float32)
+	encInt32(tag int32, x int32)
+	encSInt32(tag int32, x int32)
+	encSFixedInt32(tax int32, x int32)
+	encUInt32(tag int32, x uint32)
+	encInt64(tag int32, x int64)
+	encSInt64(tag int32, x int64)
+	encSFixedInt64(tax int32, x int64)
+	encUInt64(tag int32, x uint64)
+	encBool(tag int32, x bool)
+	encBytes(tag int32, x []byte)
+
+	// Used in cases where marshaled protobuf bytes have already been generated
+	// and need to be appended to the stream. Assumes that the <tag, wireType>
+	// tuple has already been included.
+	encPartialProto(x []byte)
+
+	bytes() []byte
+	reset()
+}
+
+type customMarshaller struct {
+	buf *buffer
+}
+
+func newCustomMarshaller() customFieldMarshaller {
+	return &customMarshaller{
+		buf: newCodedBuffer(nil),
+	}
+}
+
+func (m *customMarshaller) encFloat64(tag int32, x float64) {
+	if x == 0 {
```

Comment:
took me a minute to decipher this haha. I assumed that this would actually force the compiler to make the 0 -> 0.0 but you maybe be right that this is a type assertion. Good catch